File: query-validator-rewards.js
#!/usr/bin/env node

/**
 * Query Ethereum validator rewards using Beaconcha.in API v2
 * Usage: node query-validator-rewards.js --apikey YOUR_API_KEY [options]
 */

// Default configuration
const API_ENDPOINT = 'https://beaconcha.in/api/v2/ethereum/validators/rewards-list';
const DEFAULT_LIMIT = 100;
const DEFAULT_OFFSET = 0;

// Parse command-line arguments
function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    apikey: process.env.BEACONCHAIN_API_KEY || null,
    validators: null,
    epoch: null,
    limit: DEFAULT_LIMIT,
    offset: DEFAULT_OFFSET,
    pretty: false
  };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--apikey' || arg === '-k') {
      if (i + 1 < args.length) {
        options.apikey = args[++i];
      }
    } else if (arg === '--validators' || arg === '-v') {
      if (i + 1 < args.length) {
        // Parse comma-separated or space-separated validator IDs
        const validatorString = args[++i];
        options.validators = validatorString.split(',').map(v => parseInt(v.trim())).filter(v => !isNaN(v));
      }
    } else if (arg === '--epoch' || arg === '-e') {
      if (i + 1 < args.length) {
        options.epoch = parseInt(args[++i]);
      }
    } else if (arg === '--limit' || arg === '-l') {
      if (i + 1 < args.length) {
        options.limit = parseInt(args[++i]);
      }
    } else if (arg === '--offset' || arg === '-o') {
      if (i + 1 < args.length) {
        options.offset = parseInt(args[++i]);
      }
    } else if (arg === '--pretty' || arg === '-p') {
      options.pretty = true;
    } else if (arg === '--help' || arg === '-h') {
      printHelp();
      process.exit(0);
    }
  }

  return options;
}

function printHelp() {
  console.log(`
Beaconcha.in Validator Rewards Query Tool (JavaScript)

Usage: node query-validator-rewards.js --apikey YOUR_API_KEY [options]

Description: Query Ethereum validator rewards using Beaconcha.in API v2

Options:
  -k, --apikey <key>      Beaconcha.in API key (required)
  -v, --validators <ids>  Comma-separated list of validator indices (e.g., "1,2,3")
  -e, --epoch <epoch>     Specific epoch to query rewards for
  -l, --limit <limit>     Number of results to return (default: ${DEFAULT_LIMIT})
  -o, --offset <offset>   Pagination offset (default: ${DEFAULT_OFFSET})
  -p, --pretty            Pretty-print JSON output
  -h, --help              Show this help message

Examples:
  node query-validator-rewards.js --apikey YOUR_API_KEY
  node query-validator-rewards.js --apikey YOUR_API_KEY --validators "1,2,3" --pretty
  node query-validator-rewards.js --apikey YOUR_API_KEY --epoch 123456 --limit 50

Environment Variables:
  BEACONCHAIN_API_KEY     API key for Beaconcha.in (alternative to --apikey)

Security Note:
  Never commit API keys to version control. Use environment variables or
  gitignored configuration files. See SECURITY_BEST_PRACTICES.md for details.
`);
}

function formatResponse(data, pretty = false) {
  if (pretty) {
    return JSON.stringify(data, null, 2);
  }
  return JSON.stringify(data);
}

function formatRewardRecord(record) {
  const lines = [];
  lines.push(`  Validator Index: ${record.validatorindex ?? 'N/A'}`);
  lines.push(`  Epoch: ${record.epoch ?? 'N/A'}`);
  
  if (record.attesterslot !== null && record.attesterslot !== undefined) {
    lines.push(`  Attester Slot: ${record.attesterslot}`);
  }
  
  if (record.attestation_source !== null && record.attestation_source !== undefined) {
    lines.push(`  Attestation Source: ${record.attestation_source}`);
  }
  
  if (record.attestation_target !== null && record.attestation_target !== undefined) {
    lines.push(`  Attestation Target: ${record.attestation_target}`);
  }
  
  if (record.attestation_head !== null && record.attestation_head !== undefined) {
    lines.push(`  Attestation Head: ${record.attestation_head}`);
  }
  
  if (record.proposerslot !== null && record.proposerslot !== undefined) {
    lines.push(`  Proposer Slot: ${record.proposerslot}`);
  }
  
  if (record.deposits !== null && record.deposits !== undefined) {
    lines.push(`  Deposits: ${record.deposits}`);
  }
  
  if (record.withdrawals !== null && record.withdrawals !== undefined) {
    lines.push(`  Withdrawals: ${record.withdrawals}`);
  }
  
  return lines.join('\n');
}

async function queryValidatorRewards(options) {
  const { apikey, validators, epoch, limit, offset } = options;

  if (!apikey) {
    console.error('Error: API key is required.');
    console.error('Provide it via --apikey option or set BEACONCHAIN_API_KEY environment variable.');
    process.exit(1);
  }

  // Build request body
  const body = {
    limit: limit,
    offset: offset
  };

  if (validators && validators.length > 0) {
    body.validators = validators;
  }

  if (epoch !== null && epoch !== undefined) {
    body.epoch = epoch;
  }

  // Display query information
  console.log('Querying Beaconcha.in Validator Rewards API...');
  console.log(`Endpoint: ${API_ENDPOINT}`);
  if (validators && validators.length > 0) {
    console.log(`Validators: ${validators.join(', ')}`);
  }
  if (epoch !== null && epoch !== undefined) {
    console.log(`Epoch: ${epoch}`);
  }
  console.log(`Limit: ${limit}`);
  console.log(`Offset: ${offset}`);
  console.log();

  const fetchOptions = {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apikey}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(body)
  };

  try {
    const response = await fetch(API_ENDPOINT, fetchOptions);
    
    if (!response.ok) {
      const errorText = await response.text();
      let errorData;
      try {
        errorData = JSON.parse(errorText);
      } catch (e) {
        errorData = { error: errorText };
      }
      console.error('HTTP Error:', response.status);
      console.error('Error details:', formatResponse(errorData, true));
      process.exit(1);
    }
    
    const data = await response.json();
    
    console.log('Response:');
    console.log(formatResponse(data, options.pretty));
    
    // Check status
    const status = data.status || 'unknown';
    if (status === 'success') {
      console.log();
      console.log('✓ Query successful');
      
      // Display summary if pretty output requested
      if (options.pretty && data.data) {
        const results = data.data;
        if (Array.isArray(results)) {
          console.log(`Results: ${results.length} record(s)`);
          
          if (results.length > 0 && results.length <= 10) {
            console.log();
            console.log('Details:');
            results.forEach((record, i) => {
              console.log(`\nRecord #${i + 1}:`);
              console.log(formatRewardRecord(record));
            });
          } else if (results.length > 10) {
            console.log(`(showing first 10 of ${results.length} records)`);
            console.log();
            console.log('Details:');
            results.slice(0, 10).forEach((record, i) => {
              console.log(`\nRecord #${i + 1}:`);
              console.log(formatRewardRecord(record));
            });
          }
        }
      }
    } else if (status === 'error') {
      console.log();
      console.log('✗ Query failed');
      const errorMsg = data.error || 'Unknown error';
      console.log(`Error: ${errorMsg}`);
      process.exit(1);
    } else {
      console.log();
      console.log('⚠ Unknown response status');
    }
    
    return data;
  } catch (err) {
    console.error('Error fetching validator rewards:', err.message);
    process.exit(1);
  }
}

// Main execution
if (require.main === module) {
  const options = parseArgs();
  
  queryValidatorRewards(options)
    .then(() => {
      process.exit(0);
    })
    .catch(err => {
      console.error('Unexpected error:', err);
      process.exit(1);
    });
}

module.exports = { queryValidatorRewards };

File: query-validator-rewards.py
#!/usr/bin/env python3

"""
Beaconcha.in Validator Rewards Query Script
This script queries Ethereum validator rewards using the Beaconcha.in API v2
"""

import argparse
import json
import sys
import os
import requests

# Default configuration
API_ENDPOINT = "https://beaconcha.in/api/v2/ethereum/validators/rewards-list"
DEFAULT_LIMIT = 100
DEFAULT_OFFSET = 0


def query_validator_rewards(api_key, validators=None, epoch=None, 
                            limit=DEFAULT_LIMIT, offset=DEFAULT_OFFSET):
    """
    Query Ethereum validator rewards from Beaconcha.in API.
    
    Args:
        api_key (str): Beaconcha.in API key
        validators (list, optional): List of validator indices
        epoch (int, optional): Specific epoch to query
        limit (int): Number of results to return
        offset (int): Pagination offset
        
    Returns:
        dict: API response data
        
    Raises:
        requests.exceptions.RequestException: If API request fails
    """
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }
    
    # Build request body
    body = {
        "limit": limit,
        "offset": offset
    }
    
    if validators is not None and len(validators) > 0:
        body["validators"] = validators
    
    if epoch is not None:
        body["epoch"] = epoch
    
    try:
        response = requests.post(
            API_ENDPOINT,
            headers=headers,
            json=body,
            timeout=30
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.HTTPError as e:
        print(f"HTTP Error: {e}", file=sys.stderr)
        if e.response.text:
            try:
                error_data = e.response.json()
                print(f"Error details: {json.dumps(error_data, indent=2)}", file=sys.stderr)
            except json.JSONDecodeError:
                print(f"Response: {e.response.text}", file=sys.stderr)
        raise
    except requests.exceptions.RequestException as e:
        print(f"API request failed: {e}", file=sys.stderr)
        raise


def format_response(data, pretty=False):
    """
    Format API response data as JSON string.
    Redacts sensitive fields to prevent accidental exposure.
    
    Args:
        data: Response data to format
        pretty: Whether to pretty-print the JSON
        
    Returns:
        str: Formatted JSON string with sensitive fields redacted
    """
    import copy
    
    # List of sensitive field names to redact
    sensitive_fields = ['apikey', 'api_key', 'apiKey', 'token', 'access_token', 'secret', 'password', 'authorization']
    
    # Create a deep copy and redact sensitive fields
    sanitized = copy.deepcopy(data)
    
    def redact_sensitive_fields(obj):
        """Recursively redact sensitive fields in nested structures."""
        if not isinstance(obj, dict):
            return
        
        for key in list(obj.keys()):
            # Check if this is a sensitive field
            if any(field.lower() in key.lower() for field in sensitive_fields):
                obj[key] = '[REDACTED]'
            elif isinstance(obj[key], dict):
                # Recursively redact nested objects
                redact_sensitive_fields(obj[key])
            elif isinstance(obj[key], list):
                # Redact items in lists
                for item in obj[key]:
                    if isinstance(item, dict):
                        redact_sensitive_fields(item)
    
    redact_sensitive_fields(sanitized)
    
    if pretty:
        return json.dumps(sanitized, indent=2, sort_keys=False)
    return json.dumps(sanitized)


def format_reward_record(record):
    """Format a single reward record for display."""
    lines = []
    lines.append(f"  Validator Index: {record.get('validatorindex', 'N/A')}")
    lines.append(f"  Epoch: {record.get('epoch', 'N/A')}")
    
    if record.get('attesterslot') is not None:
        lines.append(f"  Attester Slot: {record['attesterslot']}")
    
    if record.get('attestation_source') is not None:
        lines.append(f"  Attestation Source: {record['attestation_source']}")
    
    if record.get('attestation_target') is not None:
        lines.append(f"  Attestation Target: {record['attestation_target']}")
    
    if record.get('attestation_head') is not None:
        lines.append(f"  Attestation Head: {record['attestation_head']}")
    
    if record.get('proposerslot') is not None:
        lines.append(f"  Proposer Slot: {record['proposerslot']}")
    
    if record.get('deposits') is not None:
        lines.append(f"  Deposits: {record['deposits']}")
    
    if record.get('withdrawals') is not None:
        lines.append(f"  Withdrawals: {record['withdrawals']}")
    
    return "\n".join(lines)


def main():
    """Main function to handle command-line interface and execute query."""
    parser = argparse.ArgumentParser(
        description="Query Ethereum validator rewards using Beaconcha.in API v2",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s --apikey YOUR_API_KEY
  %(prog)s --apikey YOUR_API_KEY --validators 1 2 3 --pretty
  %(prog)s --apikey YOUR_API_KEY --epoch 123456 --limit 50

Environment Variables:
  BEACONCHAIN_API_KEY    API key for Beaconcha.in (alternative to --apikey)

Security Note:
  Never commit API keys to version control. Use environment variables or
  gitignored configuration files. See SECURITY_BEST_PRACTICES.md for details.
        """
    )
    
    parser.add_argument(
        "-k", "--apikey",
        default=os.environ.get('BEACONCHAIN_API_KEY'),
        help="Beaconcha.in API key (required)"
    )
    parser.add_argument(
        "-v", "--validators",
        nargs='+',
        type=int,
        help="List of validator indices (e.g., 1 2 3)"
    )
    parser.add_argument(
        "-e", "--epoch",
        type=int,
        help="Specific epoch to query rewards for"
    )
    parser.add_argument(
        "-l", "--limit",
        type=int,
        default=DEFAULT_LIMIT,
        help=f"Number of results to return (default: {DEFAULT_LIMIT})"
    )
    parser.add_argument(
        "-o", "--offset",
        type=int,
        default=DEFAULT_OFFSET,
        help=f"Pagination offset (default: {DEFAULT_OFFSET})"
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output raw JSON response"
    )
    parser.add_argument(
        "--pretty",
        action="store_true",
        help="Pretty print output"
    )
    
    args = parser.parse_args()
    
    # Check if API key is provided
    if not args.apikey:
        print("Error: API key is required.", file=sys.stderr)
        print("Provide it via --apikey option or set BEACONCHAIN_API_KEY environment variable.", file=sys.stderr)
        sys.exit(1)
    
    # Display query information
    print("Querying Beaconcha.in Validator Rewards API...")
    print(f"Endpoint: {API_ENDPOINT}")
    if args.validators:
        print(f"Validators: {', '.join(map(str, args.validators))}")
    if args.epoch is not None:
        print(f"Epoch: {args.epoch}")
    print(f"Limit: {args.limit}")
    print(f"Offset: {args.offset}")
    print()
    
    # Execute query
    try:
        response_data = query_validator_rewards(
            args.apikey,
            validators=args.validators,
            epoch=args.epoch,
            limit=args.limit,
            offset=args.offset
        )
        
        # Output results
        print("Response:")
        if args.json or not args.pretty:
            print(format_response(response_data, pretty=args.pretty))
        else:
            print(format_response(response_data, pretty=True))
        
        # Check status
        status = response_data.get("status", "unknown")
        if status == "success":
            print()
            print("✓ Query successful")
            
            # Display summary if pretty output requested
            if args.pretty and "data" in response_data:
                results = response_data["data"]
                if isinstance(results, list):
                    print(f"Results: {len(results)} record(s)")
                    
                    if results and len(results) <= 10:
                        print()
                        print("Details:")
                        for i, record in enumerate(results, 1):
                            print(f"\nRecord #{i}:")
                            print(format_reward_record(record))
                    elif results:
                        print(f"(showing first 10 of {len(results)} records)")
                        print()
                        print("Details:")
                        for i, record in enumerate(results[:10], 1):
                            print(f"\nRecord #{i}:")
                            print(format_reward_record(record))
        elif status == "error":
            print()
            print("✗ Query failed")
            error_msg = response_data.get("error", "Unknown error")
            print(f"Error: {error_msg}")
            sys.exit(1)
        else:
            print()
            print("⚠ Unknown response status")
            
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
